examples/Parser.hs

```diff
@@ -0,0 +1,134 @@
+{-# LANGUAGE DeriveFoldable, DeriveFunctor, DeriveTraversable, ExistentialQuantification, FlexibleContexts, FlexibleInstances, KindSignatures, LambdaCase, MultiParamTypeClasses, StandaloneDeriving, TypeOperators, UndecidableInstances #-}
+module Parser
+( spec
+) where
+
+import Control.Effect
+import Control.Effect.Carrier
+import Control.Effect.Cut
+import Control.Effect.NonDet
+import Control.Effect.Sum hiding (L)
+import Control.Monad (replicateM)
+import Data.Char
+import Data.Coerce
+import Data.List (intercalate)
+import Test.Hspec
+import Test.Hspec.QuickCheck
+import Test.QuickCheck
+
+spec :: Spec
+spec = describe "parser" $ do
+  describe "parse" $ do
+    prop "returns pure values at the end of input" $
+      \ a -> run (runNonDet (parse "" (pure a))) == [a :: Integer]
+
+    prop "fails if input remains" $
+      \ c cs a -> run (runNonDet (parse (c:cs) (pure (a :: Integer)))) == []
+
+  describe "satisfy" $ do
+    prop "matches with a predicate" $
+      \ c f -> run (runNonDet (parse [c] (satisfy (applyFun f)))) == if applyFun f c then [c] else []
+
+    prop "fails at end of input" $
+      \ f -> run (runNonDet (parse "" (satisfy (applyFun f)))) == []
+
+    prop "fails if input remains" $
+      \ c1 c2 f -> run (runNonDet (parse [c1, c2] (satisfy (applyFun f)))) == []
+
+    prop "consumes input" $
+      \ c1 c2 f -> run (runNonDet (parse [c1, c2] ((,) <$> satisfy (applyFun f) <*> satisfy (applyFun f)))) == if applyFun f c1 && applyFun f c2 then [(c1, c2)] else []
+
+  describe "factor" $ do
+    prop "matches positive integers" $
+      \ a -> run (runNonDet (runCut (parse (show (abs a)) factor))) == [abs a]
+
+    prop "matches parenthesized expressions" . forAll (sized arbNested) $
+      \ as -> run (runNonDet (runCut (parse ('(' : intercalate "+" (intercalate "*" . map (show . abs) . (1:) <$> [0]:as) ++ ")") factor))) == [sum (map (product . map abs) as)]
+
+  describe "term" $ do
+    prop "matches factors" $
+      \ a -> run (runNonDet (runCut (parse (show (abs a)) term))) == [abs a]
+
+    prop "matches multiplication" $
+      \ as -> run (runNonDet (runCut (parse (intercalate "*" (show . abs <$> 1:as)) term))) == [product (map abs as)]
+
+  describe "expr" $ do
+    prop "matches factors" $
+      \ a -> run (runNonDet (runCut (parse (show (abs a)) expr))) == [abs a]
+
+    prop "matches multiplication" $
+      \ as -> run (runNonDet (runCut (parse (intercalate "*" (show . abs <$> 1:as)) expr))) == [product (map abs as)]
+
+    prop "matches addition" $
+      \ as -> run (runNonDet (runCut (parse (intercalate "+" (show . abs <$> 0:as)) expr))) == [sum (map abs as)]
+
+    prop "respects order of operations" . forAll (sized arbNested) $
+      \ as -> run (runNonDet (runCut (parse (intercalate "+" (intercalate "*" . map (show . abs) . (1:) <$> [0]:as)) expr))) == [sum (map (product . map abs) as)]
```

Comment:
This doesn’t produce arbitrarily-nested expressions, merely sums of products, but it’s still an interesting test.